crates/wast/src/ast/module.rs

```diff
@@ -36,9 +37,45 @@ impl<'a> Parse<'a> for Wat<'a> {
     }
 }
 
+/// A `*.cat` file parser, or a parser for one parenthesized component module.
+///
+/// This is the top-level type which you'll frequently parse when working with
+/// this crate. A `*.cat` file is either one `component` s-expression or a
+/// sequence of s-expressions that are module fields.
+#[derive(Debug)]
+pub struct Cat<'a> {
+    #[allow(missing_docs)]
+    pub module: Module<'a>,
+}
+
+impl<'a> Parse<'a> for Cat<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        if !parser.has_meaningful_tokens() {
+            return Err(parser.error("expected at least one component field"));
+        }
+        let _r = parser.register_annotation("custom");
+        let module = if !parser.peek2::<kw::component>() {
+            let fields = ModuleField::parse_remaining(parser)?;
+            Module {
+                is_component: true,
+                span: ast::Span { offset: 0 },
+                id: None,
+                name: None,
+                kind: ModuleKind::Text(fields),
+            }
+        } else {
+            parser.parens(|parser| parser.parse())?
+        };
+        module.validate(parser)?;
+        Ok(Cat { module })
+    }
+}
+
 /// A parsed WebAssembly module.
 #[derive(Debug)]
 pub struct Module<'a> {
+    /// Is this a component module, rather than a core module?
+    pub is_component: bool,
```

Comment:
One possible downside to this implementation is that this may run the risk of making the internals more complicated than they otherwise need to be or make encoding more difficult. For example a components and modules have a sort of a venn-diagram of supported fields. For example exports/imports are pretty similar (but imports are notably different with one-or-two levels) but otherwise modules allow for table definitions while components only allow for aliases. Similarly data segments and element segments are allowed in modules but not components.

This probably mostly adds complexity to the resolver in the `wast` crate. Today it handles nested modules as if they could contain everything because that's what module linking originally was, but we actually in theory should be able to remove a ton of that complexity since we modules are always leaves and components can only refer, in a limited fashion, to other items.

Overall I'm not sure if this is the best representation or whether it would be better to have this split (a `Module` type and a `Component` type). You mentioned that having `Component` might introduce a good deal of duplication but I feel like a lot of complexity, especially around the resolver, may end up getting shifted largely to components in such a design. 

I wanted to write this down, but again I'm not sure what the right answer is. It's probably best to just pick one you find reasonable and go with that but I think we should keep our eyes peeled for places where it may be better to go with an alternate design.